### PR TITLE
cs/fix/bulk async zip

### PR DESCRIPTION
**Pull request by chancesmith**
## overall purpose
Need to ensure all files and content is ready before zipping up.

## changes made
- add async to file content collection

## screenshot
Screen Shot 2022-12-30 at 3 00 24 PM

### CODE REVIEW

File: csv-bulk-signature/src/main.ts [hunk 1]
```diff
@@ -25,22 +34,6 @@ interface TemplateData {
...
-const SIGNATURES_PATH = "./dist/signatures";
-const LOGOS = {
-  "ata-cpa-advisors":
-    "https://temp-ata-signature-assets.s3.amazonaws.com/ATA_LOGO-CPAAdvisor-BT-RGB.png",
-  "ata-capital":
-    "https://temp-ata-signature-assets.s3.amazonaws.com/ATAC_LOGO-BT-RGB.png",
-  "ata-employment-solutions":
-    "https://temp-ata-signature-assets.s3.amazonaws.com/ATAES_LOGO-BT-RGB.png",
-} as const;
-
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:43]: move globals into config file (trim file length)

File: csv-bulk-signature/src/main.ts [hunk 2]
```diff
@@ -69,33 +62,22 @@ async function generateSignatures(
...
-    const logoId = contact["Brand*"];
-    const logoUrl = LOGOS[logoId];
-    const fullName = contact["Full Name*"];
-    const credentials = contact["Credentials"];
-    const title = contact["Title*"];
-    const officePhone = contact["Office Phone*"];
-    const mobilePhone = contact["Mobile Phone"];
-    const calendly = contact["Calendly Link"];
+  for await (const contact of contacts) {
+    if (contact.skip) return; // skip if missing required fields
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:66]: use for/await/of for async looping

File: csv-bulk-signature/src/main.ts [hunk 3]
```diff
@@ -69,33 +62,22 @@ async function generateSignatures(
...
-      mobilePhone,
-      calendly,
+      logoUrl: LOGOS[contact["Brand*"]],
+      fullName: contact["Full Name*"],
+      credentials: contact["Credentials"],
+      title: contact["Title*"],
+      officePhone: contact["Office Phone*"],
+      mobilePhone: contact["Mobile Phone"],
+      calendly: contact["Calendly Link"],
     });
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:76]: inline variables because we're only using most of them once

File: csv-bulk-signature/src/main.ts [hunk 4]
```diff
@@ -69,33 +62,22 @@ async function generateSignatures(
...
 
-    await fs.promises.writeFile(
-      `${SIGNATURES_PATH}/${fullName.split(" ").join("")}.htm`,
-      html
-    );
-  });
+    const fileName = contact["Full Name*"].split(" ").join("");
+    await fs.promises.writeFile(`${SIGNATURES_PATH}/${fileName}.htm`, html);
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:79]: concise 4 lines to 2 (enjoying the trimming a bit too much)

File: csv-bulk-signature/src/main.ts [hunk 5]
```diff
@@ -124,57 +106,51 @@ async function checkHeadersToBeSame(contact: Contact) {
   if (hasExpectedHeaders) {
     console.log("👍 CSV Headers match");
   } else {
-    throw new Error("🙅‍♂️ CSV Headers do not match");
+    throw new Error("🔴 CSV Headers do not match");
   }
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:110]: make it clearer to the developer something is wrong (RED)

File: csv-bulk-signature/src/main.ts [hunk 6]
```diff
@@ -124,57 +106,51 @@ async function checkHeadersToBeSame(contact: Contact) {
...
 }
 
 async function zipUpFile() {
-  const signatureFolder = fs.readdirSync(SIGNATURES_PATH);
+  const signatureFolder = await fs.promises.readdir(SIGNATURES_PATH);
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts]: get folder with a promise

File: csv-bulk-signature/src/main.ts [hunk 7]
```diff
@@ -124,57 +106,51 @@ async function checkHeadersToBeSame(contact: Contact) {
...
-      zip.file(file, fileContents);
-    } catch (error: any) {
-      console.error(`Error adding file ${file}: ${error.message}`);
-    }
-  });
+  for await (const file of files) {
+    const filePath = `${SIGNATURES_PATH}/${file}`;
+    const fileContents = await fs.promises.readFile(filePath);
+    zip.file(file, fileContents);
+  }
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts]: use for/await/of for async looping

File: csv-bulk-signature/src/main.ts [hunk 8]
```diff
@@ -124,57 +106,51 @@ async function checkHeadersToBeSame(contact: Contact) {
...
+    const buffer = await zip.generateAsync({ type: "nodebuffer" });
 
-  await fs.promises.writeFile(
-    `${SIGNATURES_PATH}-${
-      new Date().toISOString().split(":").join("_").split(".")[0] // date and time
-    }.zip`,
-    buffer
-  );
-  console.log("👍 Zip file created (signatures + report)");
+    await fs.promises.writeFile(ZIP_FILE, buffer);
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts]: change zip file to const in config since used twice in app

File: csv-bulk-signature/src/main.ts [hunk 9]
```diff
@@ -124,57 +106,51 @@ async function checkHeadersToBeSame(contact: Contact) {
...
-    const hasRequiredData = checkRequiredFields(contact);
-    if (!hasRequiredData) {
+  for (const contact of contacts) {
+    if (contact.skip) {
       skippedRows.push(contact["Full Name*"]);
     } else {
       processedRows.push(contact["Full Name*"]);
     }
-  });
+  }
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:149]: chance foreach (better for dom manipulation) to for loop (modern)

File: csv-bulk-signature/src/main.ts [hunk 10]
```diff
@@ -187,10 +163,7 @@ async function createStatusReport(contacts: Contact[]) {
     `\nROWS/SIGNATURES PROCESSED SUCCESSFULLY: (${processedRows.length}) \n`
   );
   const combinedStatus = skippedRows.concat(processedRows);
-  await fs.promises.writeFile(
-    `${SIGNATURES_PATH}/_statusReport.txt`,
-    combinedStatus.join("\n")
-  );
+  await fs.promises.writeFile(STATUS_REPORT_FILE, combinedStatus.join("\n"));
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:166]: make report file name a const in config too

File: csv-bulk-signature/src/main.ts [hunk 11]
```diff
@@ -141,44 +118,39 @@ async function zipUpFile() {
...
-      console.error(`Error adding file ${file}: ${error.message}`);
-    }
-  });
-
-  const buffer = await zip.generateAsync({ type: "nodebuffer" });
+  for await (const file of files) {
+    const filePath = `${SIGNATURES_PATH}/${file}`;
+    const fileContents = await fs.promises.readFile(filePath);
+    zip.file(file, fileContents);
+  }
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:125]: use for/await/of for async looping

File: csv-bulk-signature/src/main.ts [hunk 12]
```diff
@@ -141,44 +118,39 @@ async function zipUpFile() {
...
 
-  const zipFilePath = `${SIGNATURES_PATH}-${
-    new Date().toISOString().split(":").join("_").split(".")[0] // date and time
-  }.zip`;
+  try {
+    const buffer = await zip.generateAsync({ type: "nodebuffer" });
 
-  await fs.promises.writeFile(zipFilePath, buffer);
-  console.log("👍 Zip file created (signatures + report)");
+    await fs.promises.writeFile(ZIP_FILE, buffer);
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:130]: change zip file to const in config since used twice in app

File: csv-bulk-signature/src/main.ts [hunk 13]
```diff
@@ -141,44 +118,39 @@ async function zipUpFile() {
...
 async function createStatusReport(contacts: Contact[]) {
   const skippedRows: string[] = [];
   const processedRows: string[] = [];
 
   // 1. collect skipped and processed rows
-  contacts.forEach((contact, index) => {
-    const hasRequiredData = checkRequiredFields(contact);
-    if (!hasRequiredData) {
+  for (const contact of contacts) {
+    if (contact.skip) {
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:144]: change foreach (good for DOM manipulation) for a for loop

File: csv-bulk-signature/src/main.ts [hunk 14]
```diff
@@ -278,8 +247,8 @@ async function main() {
   // 4. generate signatures + report + zip
   await generateSignatures(template, contactsWithNewProps);
   await createStatusReport(contactsWithNewProps);
-  const zipFilePath = await zipUpFile();
-  await checkZipIsNotEmpty(zipFilePath);
+  await zipUpFile();
+  await checkZipIsNotEmpty(ZIP_FILE);
```
Comment (chancesmith) [inline csv-bulk-signature/src/main.ts:251]: allowed zip file name to be consistent across the board by creating it as a const

**Review by adamnitti — APPROVED** [review]

**Pull request merged**